Implement cancel event on <input type=file>

rr-codes opened the pull request:
#### 059f8069cb4a83b08a89b8b2ae02d9b82d8c5fd1
<pre>
Implement cancel event on &lt;input type=file&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=227799">https://bugs.webkit.org/show_bug.cgi?id=227799</a>
rdar://80340803

Reviewed by Aditya Keerthi.

This PR implements the &quot;Add a &quot;cancel&quot; event for when file upload selection is unchanged&quot;
behavior.

The `cancel` event is now fired when:
- The file picker is dismissed
- The selected files are the same as those currently selected

Spec: <a href="https://html.spec.whatwg.org/multipage/input.html#show-the-picker">https://html.spec.whatwg.org/multipage/input.html#show-the-picker</a>,-if-applicable

* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::setFiles):
(WebCore::FileInputType::fileChoosingCancelled):
* Source/WebCore/html/FileInputType.h:
* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::cancelFileChoosing):
* Source/WebCore/platform/FileChooser.h:
* Source/WebKit/WebProcess/WebPage/WebOpenPanelResultListener.cpp:
(WebKit::WebOpenPanelResultListener::didCancelFileChoosing):
* Source/WebKit/WebProcess/WebPage/WebOpenPanelResultListener.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didCancelForOpenPanel):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
(-[FileInputTypeCancelEventUIDelegate webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:]):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/255394@main">https://commits.webkit.org/255394@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f4e6090fd81108588de31cf60dfc6da2e9125ced

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92081 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1313 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22594 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101874 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161929 "Reverted pull request changes (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96081 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1310 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29711 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84523 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98050 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97739 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/820 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78606 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27766 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82742 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82341 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70809 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36137 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16355 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33889 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17457 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3737 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37760 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40166 "Found 2 new test failures: imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.html?h264_annexb, imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.worker.html?h264_annexb (failure)") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39657 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36583 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
This needs a test.

pxlcoder commented on Source/WebCore/html/FileInputType.cpp:
```diff
@@ -400,6 +400,8 @@ void FileInputType::setFiles(RefPtr<FileList>&& files, RequestIcon shouldRequest
         protectedInputElement->dispatchChangeEvent();
+    } else {
+        protectedInputElement->dispatchCancelEvent();
     }
```
Style: Remove braces on single-line `else` statement.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
```diff
@@ -79,6 +93,43 @@ - (void)webView:(WKWebView *)webView runOpenPanelWithParameters:(WKOpenPanelPara
+    NSPoint clickPoint = NSMakePoint(50, 50);
+    NSInteger windowNumber = [webView window].windowNumber;
+    [webView mouseDown:[NSEvent mouseEventWithType:NSEventTypeLeftMouseDown location:clickPoint modifierFlags:0 timestamp:0 windowNumber:windowNumber context:nil eventNumber:0 clickCount:1 pressure:1]];
+    [webView mouseUp:[NSEvent mouseEventWithType:NSEventTypeLeftMouseUp location:clickPoint modifierFlags:0 timestamp:0 windowNumber:windowNumber context:nil eventNumber:0 clickCount:1 pressure:1]];
```
`sendClickAtPoint`

rr-codes replied:
@pxlcoder what does this comment mean?

pxlcoder replied:
There's a `sendClickAtPoint` method which would simplify this logic.

pxlcoder commented on Source/WebCore/platform/FileChooser.cpp:
```diff
@@ -72,16 +68,17 @@ void FileChooser::chooseFiles(const Vector<String>& filenames, const Vector<Stri
 
+void FileChooser::cancelFileChoosing()
+{
+    m_client->fileChoosingCancelled();
```
Null-check `m_client` like the other methods here?

rr-codes committed:
Implement cancel event on <input type=file>
https://bugs.webkit.org/show_bug.cgi?id=227799
rdar://80340803

Reviewed by Aditya Keerthi.

This PR implements the "Add a "cancel" event for when file upload selection is unchanged"
behavior.

The `cancel` event is now fired when:
- The file picker is dismissed
- The selected files are the same as those currently selected

Spec: https://html.spec.whatwg.org/multipage/input.html#show-the-picker,-if-applicable

* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::setFiles):
(WebCore::FileInputType::fileChoosingCancelled):
* Source/WebCore/html/FileInputType.h:
* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::cancelFileChoosing):
* Source/WebCore/platform/FileChooser.h:
* Source/WebKit/WebProcess/WebPage/WebOpenPanelResultListener.cpp:
(WebKit::WebOpenPanelResultListener::didCancelFileChoosing):
* Source/WebKit/WebProcess/WebPage/WebOpenPanelResultListener.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didCancelForOpenPanel):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
(-[FileInputTypeCancelEventUIDelegate webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:]):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/255394@main
